alpha/ui.cpp

```diff
@@ -31,69 +31,157 @@ void _TS_UI::begin()
+    clickP = this->read_button(TS_HAL.btn_power_get() != TS_ButtonState::NotPressed, btnP);
+
+    if (clickA) {
+      if (selected == cursor) {
+        selected = -1;
+      } else {
+        selected = cursor;
+      }
+    }
+    if (selected == 0) {
+        options[0][brightness / 20 + 11] = '-';
+      if (clickB) {
+        brightness %= 100;
+        brightness += 20;
+      }
+      if (clickP) {
+        brightness += 60;
+        brightness %= 100;
+        brightness += 20;
+      }
+      options[0][brightness / 20 + 11] = '|';
+      TS_HAL.lcd_brightness(brightness);
+    } else {
+      if (clickB) {
+        ++cursor %= 3;
+        selected = -1;
+      }
+      if (clickP) {
+        cursor += 2;
+        cursor %= 3;
+        selected = -1;
+      }
+    }
+
+    bool hasUpdate = clickA || clickB || clickP;
+
+    if (hasUpdate) {
+      if (millis() - savePowerStart < MIN_SLEEP_DURATION) {
+        continue;
+      }
+      if (savePowerStart) {
+        savePowerStart = 0;
+        selected = -1;
+        TS_HAL.lcd_sleep(false);
+        TS_HAL.lcd_brightness(brightness);
+      }
 
-  while(true)
-  {
-    // don't update contents if lcd is not turned on
-    if(!SavePower)
-    {
       TS_DateTime datetime;
       TS_HAL.rtc_get(datetime);
+
+      TS_HAL.lcd_cursor(0, 0);
+      // TODO: Does not work with F()
+
+      TS_HAL.lcd_printf("%04d-%02d-%02d ", datetime.year, datetime.month, datetime.day);
+      TS_HAL.lcd_printf("%02d:%02d:%02d\n", datetime.hour, datetime.minute, datetime.second);
+      TS_HAL.lcd_printf("Battery: %d%%", TS_HAL.power_get_batt_level());
+      if (TS_HAL.power_is_charging()) {
+        TS_HAL.lcd_printf(", Charging");
+      }
+      TS_HAL.lcd_printf("\n");
       
-      TS_HAL.lcd_cursor(0, 15);
-      // TODO: does not work with F()
-      TS_HAL.lcd_printf("Date: %04d-%02d-%02d\n",     datetime.year, datetime.month, datetime.day);
-      TS_HAL.lcd_printf("Time: %02d : %02d : %02d\n", datetime.hour, datetime.minute, datetime.second);
-      TS_HAL.lcd_printf("Battery: %d%%  ", TS_HAL.power_get_batt_level(), NULL, NULL);
-      if (TS_HAL.power_is_charging()) 
-      {
-        TS_HAL.lcd_printf("Status: Charging    ");
-      } else {
-        TS_HAL.lcd_printf("Status: Not Charging");
+      for (int i = 0; i < 3; i++) {
+        if (selected == i) {
+          TS_HAL.lcd_printf(" [%s]  \n", options[i]);
+        } else if (cursor == i) {
+          TS_HAL.lcd_printf("> %s   \n", options[i]);
+        } else {
+          TS_HAL.lcd_printf("  %s   \n", options[i]);
+        }
       }
-    }
 
-    // stackHighWaterMark = uxTaskGetStackHighWaterMark(NULL);
-    // Serial.print("UI: ");
-    // Serial.println(stackHighWaterMark);
+      stackHighWaterMark = uxTaskGetStackHighWaterMark(NULL);
+      Serial.print("UI: ");
+      Serial.println(stackHighWaterMark);
+
+      // Select option 3 to sleep.
+      if (selected == 2) {
```

Comment:
shouldn't the device sleep itself after some time of idle? or do you mean deep-sleep / shutoff

Reply:
not implemented yet